### ISSUE TITLE
Exporting arrays of hints

### ISSUE THREAD

**Issue by bojidar-bg** [godotengine/godot#3586]
**TL;DR;** Support exporting of arrays with a given export hint, as in `export([], String, FILE)`

---

It would be nice if one can export an array of some export type, like an array of enums, an array of filepaths, etc. So, in other words, support not only:

``` gdscript
export(String, FILE) var file
export(NodePath) var nodepath
```

but also:

``` gdscript
export([], String, FILE) var files
export([], NodePath) var nodepaths
```

or even:

``` gdscript
export([], [], String, FILE) var files_2d_array
export([], [], NodePath) var nodepaths_2d_array
```

---

Because syntax needs some discussion, I would go on and propose a few variants:
- 1a. Square brackets (`[` and `]`) before other export parameters

``` gdscript
export([], String, FILE) var files
export([], [], float) var heights
export([], Array) var anything_array_2d
export([], []) var anything_array_2d # Or maybe not?
```
- 1b. Array before other export parameters

``` gdscript
export(Array, String, FILE) var files
export(Array, Array, float) var heights
export(Array, Array) var anything_array_2d
```
- 2a. Brackets after export parameters

``` gdscript
export(String, FILE)[] var files
export(float)[][] var heights
export[][] var anything_array_2d
```
- 2b. Brackets after export parameters, alteranative multi-dimetional

``` gdscript
export(float)[,] var heights
export[,] var anything_array_2d
```
- 3a. Brackets before export parameters

``` gdscript
export[](String, FILE) var files
export[][](float) var heights
export[][] var anything_array_2d # Same as 2a
```
- 3b. Brackets before export parameters, alteranative multi-dimetional

``` gdscript
export[,](float) var heights
export[,] var anything_array_2d # Same as 2b
```
- 4a. Plural inflections on params, disallowing multi-dimensional exports

``` gdscript
export(String, FILE)s var files
export(Array)s var anything_array_2d
# Can't export typed multidimenstional arrays, such as float[][]
```
- 4b. Plural inflections on keyword, disallowing multi-dimensional exports

``` gdscript
exports(String, FILE) var files
exports(Array) var anything_array_2d
# Can't export typed multidimenstional arrays, such as float[][]
```

As for my own preference, I like `1b` and `3a` the most, because I think they would match the current style. But `1a` isn't too bad either. I wouldn't go with any of `2b`, `3b`, `4a` or `4b`, as they don't match the style well enough.

---

P.S. I'm not looking for workarounds, and I know that they exist for `NodePaths` or 2D arrays. Also, some things like exported enums or file paths can't be really worked around without editor plugins for the moment.

P.P.S. There was a similar issue earlier #3388, but it wasn't as general as this one, and it got closed as it was asking for some way to assign resources (not any export type) in arrays.

**Comment by LeonardMeagher2** [comment]
It'd also be nice if you could export a subClass with exported variables exposed, and also export an array of subclasses.

**Comment by bojidar-bg** [comment]
@LeonardMeagher2 Better open another issue about it, as this is about arrays, and your is about custom types.

**Comment by reduz** [comment]
Typed arrays are planned, in fact.. but not for the near future. Tagging 3.0 for now

**Comment by bojidar-bg** [comment]
@reduz I'm not talking about typed arrays, but about exporting arrays of hints.

**Comment by akien-mga** [comment]
Not critical for the upcoming 2.1, so moving to the next milestone.

**Comment by henriiquecampos** [comment]
This would made plugins and tools so much more easier. I'm now trying to make an arrays of meshes, and the only way currently is to set a general purpose Array, then create an index of type Object, then set the Object to be a Mesh. Too many steps for, let's say, an artist if he/she want to set 5 Meshes with 5 LOD.

**Comment by bojidar-bg** [comment]
Hm.. guess I can try doing it :smiley:
